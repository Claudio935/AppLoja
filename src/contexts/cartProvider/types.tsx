export enum types {
  TYPE_ADD_PRODUCT = "TYPE_ADD_PRODUCT",
  TYPE_DELETE_CART = "TYPE_DELETE_CART",
  TYPE_INCREMENT_CART = "TYPE_INCREMENT_CART",
  TYPE_DECREMENT_CART = "TYPE_DECREMENT_CART",
  TYPE_ADD_CART = "TYPE_ADD_CART",
  TYPE_CLEAN_CART = "TYPE_CLEAN_CART",
}