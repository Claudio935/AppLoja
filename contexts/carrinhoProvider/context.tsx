import React,{ createContext } from 'react'
import { Action, PropsCarrinho, PropsCarrinhoState } from '../../types/interfaces';









export const defaultValue: PropsCarrinhoState =  {
  carrinho:[
  ]}


export const CarrinhoContext = createContext<PropsCarrinhoState>(defaultValue)