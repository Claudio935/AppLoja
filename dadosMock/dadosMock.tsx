export const mockCarrinho = [{
    title: 'testMasculino',
    image: './img/testMasculino',
    price: 1.5,
    description: 'testandoMasculino',
    id: '1',
    category: "men's clothing"
  
  },
  {
    title: 'testFeminino',
    image: './img/testFeminino',
    price: 2,
    description: 'testandoFeminino',
    id: '2',
    category: "women's clothing"
  
  },
  {
    title: 'testEletronico',
    image: './img/testEletronico',
    price: 3,
    description: 'testandoEletronico',
    id: '3',
    category: "electronics"
  
  },
  
  {
    title: 'testJoia',
    image: './img/testJoia',
    price: 4,
    description: 'testandoJoia',
    id: '4',
    category: "jewelery"
  
  }
  ]