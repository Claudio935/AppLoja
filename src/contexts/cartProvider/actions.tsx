import { PropsCart, type Action, } from "../../types/types"
import { types } from "./types"

export const addProduto = (dispatch: React.Dispatch<Action>, payload: PropsCart): void => {
  dispatch({ type: types.TYPE_ADD_CART, payload })
}

export const deleteProduct = (dispatch: React.Dispatch<Action>, payload: PropsCart): void => {
  dispatch({ type: types.TYPE_DELETE_CART, payload })
}
export const IncrementProduct = (dispatch: React.Dispatch<Action>, payload: PropsCart): void => {
  dispatch({ type: types.TYPE_INCREMENT_CART, payload })
}
export const decrementProduct = (dispatch: React.Dispatch<Action>, payload: PropsCart): void => {
  dispatch({ type: types.TYPE_DECREMENT_CART, payload })
}

