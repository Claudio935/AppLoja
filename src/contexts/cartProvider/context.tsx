import { createContext } from 'react'
import { type Action, type PropsCartState } from '../../types/types';


export const defaultValue: PropsCartState = {
  cart: []
}


export const CartContext = createContext<{
  store: PropsCartState;
  dispatch: React.Dispatch<Action>;
}>({
  store: defaultValue,
  dispatch: () => null
})