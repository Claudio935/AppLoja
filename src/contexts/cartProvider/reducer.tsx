import { type Action, type PropsCartState } from "../../types/types";
import { types } from "./types";

export const reducer = (store: PropsCartState, action: Action): PropsCartState => {

  const { cart } = store


  switch (action.type) {
    case types.TYPE_ADD_CART:


      return { cart: [...cart, action.payload] };

    case types.TYPE_DELETE_CART:


      return { cart: cart.filter((item) => item.id !== action.payload.id) }

    case types.TYPE_INCREMENT_CART:

      cart.forEach((product, index) => {
        if (product.id === action.payload.id) {
          const newCart = cart.splice(index, 1, action.payload)
          return { cart: newCart }
        }
      })

      return { cart }
    case types.TYPE_DECREMENT_CART:

      cart.forEach((product, index) => {

        if (product.id === action.payload.id) {
          const newCart = cart.splice(index, 1, action.payload)

          return { cart: newCart }
        }
      })


      return { cart }

    case "RELOAD":


      return { cart: action.payload };

    default:
      return store;

  }

};
