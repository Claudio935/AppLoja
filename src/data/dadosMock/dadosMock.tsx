export const mockCart = {
  cart: [{
    title: 'testMasculino',
    image: './img/testMasculino',
    price: 1.5,
    description: 'testandoMasculino',
    id: '1',
    category: "men's clothing",
    quantidade: 1,
  },
  {
    title: 'testFeminino',
    image: './img/testFeminino',
    price: 2,
    description: 'testandoFeminino',
    id: '2',
    category: "women's clothing",
    quantidade: 1,
  },
  {
    title: 'testEletronico',
    image: './img/testEletronico',
    price: 3,
    description: 'testandoEletronico',
    id: '3',
    category: "electronics",
    quantidade: 1,
  },

  {
    title: 'testJoia',
    image: './img/testJoia',
    price: 4,
    description: 'testandoJoia',
    id: '4',
    category: "jewelery",
    quantidade: 1,

  }
  ],
  dispatch: () => null
}

export const mockUseContext = {
  store: {
    cart:
      [
        {
          title: 'testMasculino',
          image: './img/testMasculino',
          price: 1.5,
          description: 'testandoMasculino',
          id: '102',
          category: "men's clothing",
          quantidade: 1,
        },
      ],
  },

  dispatch: jest.fn()
}