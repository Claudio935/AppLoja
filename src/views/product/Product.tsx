import { RouteProp, useNavigation, useRoute } from '@react-navigation/native';
import React, { useState, useContext, useEffect } from 'react';
import {
  Image,
  StyleSheet,
  Text,
  View,
  Button,
  ScrollView,
  Alert
} from 'react-native';
import { CartContext } from '../../contexts/cartProvider/context';
import { addProduto } from '../../contexts/cartProvider/actions';
import { PropsCart, RootStackParamList } from '../../types/types';
import { StackNavigation } from '../../../App';
import { transformMoney } from '../../utils/transformFunctions';


type ProfileScreenRouteProp = RouteProp<RootStackParamList, 'Produto'>;

function Product(): JSX.Element {

  const { params } = useRoute<ProfileScreenRouteProp>();

  const navigation = useNavigation<StackNavigation>();
  const { store, dispatch } = useContext(CartContext);

  const [quantidadeProduto, setQuantidadeProduto] = useState(0);
  const [objectCarrinho, setObjectCarrinho] = useState<PropsCart>({
    title: "",
    image: "",
    quantidade: quantidadeProduto,
    price: 0,
    id: '',
    description: '',
    category: '',
  });

  useEffect(() => {
    if (params) {
      setObjectCarrinho({
        title: params.title,
        image: params.image,
        quantidade: quantidadeProduto,
        price: params.price,
        id: params.id,
        description: params.description,
        category: params.description
      })
    }

  }, [params, quantidadeProduto]);


  const handleAddCarrinho = (item: PropsCart) => {
    if (quantidadeProduto === 0) {
      Alert.alert(
        "Alerta",
        "Adicione uma quantidade de produto",
        [
          {
            text: "Cancelar",
            style: "cancel"
          },

        ]
      );
      return
    };

    if (store.cart.filter((objeto) => objeto.id === params?.id).length > 0) {
      Alert.alert(
        "Alerta",
        "Produto j√° adicionado?",
        [
          {
            text: "Cancelar",
            style: "cancel"
          },

        ]
      );
      return
    }

    addProduto(dispatch, {
      title: item.title,
      image: item.image,
      quantidade: quantidadeProduto,
      price: item.price,
      id: item.id,
      description: item.description,
      category: item.category
    })

    navigation.navigate('Cart')
  };


  return (
    <ScrollView>
      <View style={styles.container}>
        <View style={styles.produtoContainer}>
          <Image source={{ uri: params?.image }} style={styles.produtoImage} testID='imageProduto'></Image>
          <Text style={styles.titleProduto}>{params?.title}</Text>
          <Text style={styles.description}>{params?.description}</Text>
          <Text style={styles.price}>{transformMoney(!params?.price ? 0 : params.price * quantidadeProduto)}</Text>
          <View style={styles.quantifyContainer}>
            <Button title='<' onPress={() => { setQuantidadeProduto((value) => value === 0 ? value : value - 1); }} testID='decrement'></Button>
            <Text style={styles.textButton} testID='textQuantidade'>{quantidadeProduto}</Text>
            <Button title='>' onPress={() => { setQuantidadeProduto((value) => value + 1); }} testID='increment'></Button>
          </View>
          <Button title='Adicionar ao Carrinho' onPress={() => { handleAddCarrinho(objectCarrinho); }} testID='carrinhoButton'></Button>
        </View>
      </View>
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'column',
    flex: 1,
    marginBottom: 15,
  },
  produtoContainer: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    flexDirection: 'column',
    marginVertical: 15,
    width: '100%',

  },
  produtoImage: {
    width: 180,
    height: 220,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 15,
  },
  quantifyContainer: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
    width: '100%',
    marginBottom: 15,
  },
  textButton: {
    marginHorizontal: 15,
    fontWeight: '900'
  },

  titleProduto: {
    fontSize: 15,
    fontWeight: '900',
    color: '#000',
    width: 300,
    textAlign: 'center',
    marginBottom: 15,
  },
  description: {
    textAlign: 'justify',
    paddingHorizontal: 5,
    marginBottom: 15,
  },
  price: {
    fontSize: 14,
    color: '#b71919',
    fontWeight: '900',
    marginBottom: 15,
  }

});


export default Product;

