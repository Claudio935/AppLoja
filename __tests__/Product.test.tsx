import { render, screen, fireEvent, cleanup } from "@testing-library/react-native";
import { NavigationContainer } from "@react-navigation/native";
import "@testing-library/jest-dom";
import React from "react";
import { Alert } from "react-native";
import renderer from 'react-test-renderer';
import Product from "../src/views/product/Product";

import { CartContext } from "../src/contexts/cartProvider/context";
import { mockCart } from "../src/data/dadosMock/dadosMock";
import { transformMoney } from "../src/utils/transformFunctions";


const mockUseContext = {
  store: {
    cart:
      [
        mockCart.cart[0]
      ],
  },

  dispatch: jest.fn()
}

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useRoute: () => ({
    params: mockCart.cart[0],
  }),
}));

jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');
  RN.Alert.alert = jest.fn();
  return RN;
});

describe("<Product />", () => {
  beforeEach(() => {
    render(
      <CartContext.Provider value={mockUseContext}>
        <NavigationContainer>
          <Product />
        </NavigationContainer>
      </CartContext.Provider>

    );
  });
  it('renders correctly', () => {


    const tree = renderer
      .create(
        <NavigationContainer>
          <Product />
        </NavigationContainer>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should render and show components ", async () => {

    // testando a aparição do produto
    const imgProduto = screen.getByTestId('imageProduto');
    const titleProduto = screen.getByText(`${mockCart.cart[0].title}`)
    const descriptionProduto = screen.getByText(`${mockCart.cart[0].description}`);


    expect(imgProduto).toBeTruthy();
    expect(titleProduto).toBeTruthy();
    expect(descriptionProduto).toBeTruthy();
    expect(imgProduto.props.source.uri).toEqual(`${mockCart.cart[0].image}`);

    // testando a existencia dos componentes relacionado á quantidade do produto
    const buttonQuantidadeIncrement = screen.getByTestId("increment");
    const buttonQuantidadeDecrement = screen.getByTestId("decrement");
    const textButtonQuantidadeIncrement = screen.getByText(">");
    const textButtonQuantidadeDecrement = screen.getByText("<");


    expect(textButtonQuantidadeIncrement).toBeTruthy();
    expect(textButtonQuantidadeDecrement).toBeTruthy();

    expect(buttonQuantidadeIncrement).toBeTruthy();
    expect(buttonQuantidadeDecrement).toBeTruthy();

    // testando a existencia do botão de carrinho
    const texButton = screen.getByText('Adicionar ao Carrinho');
    expect(texButton).toBeTruthy();


  });

  it("changes quantity and price correctly.", async () => {


    const useContextSpy = jest.spyOn(React, 'useContext');
    useContextSpy.mockReturnValue(mockUseContext);

    // testando adição de quantidade e a mudança do preço
    const textQuantidade = screen.getByTestId('textQuantidade');

    const buttonQuantidadeIncrement = screen.getByTestId("increment");
    const buttonQuantidadeDecrement = screen.getByTestId("decrement");


    fireEvent.press(buttonQuantidadeIncrement);
    fireEvent.press(buttonQuantidadeIncrement);

    expect(textQuantidade.props.children).toEqual(2);
    fireEvent.press(buttonQuantidadeDecrement);
    expect(textQuantidade.props.children).toEqual(1);


    const textPrice = screen.getByText(transformMoney(mockCart.cart[0].price));

    expect(textPrice).toBeTruthy();

    fireEvent.press(buttonQuantidadeDecrement);
    fireEvent.press(buttonQuantidadeDecrement);
    expect(textQuantidade.props.children).toEqual(0);


  });
  it("Test the quantity and price change action of an item on button click and show alert id", async () => {


    const buttonQuantidadeIncrement = screen.getByTestId("increment");
    fireEvent.press(buttonQuantidadeIncrement);

    // testando o alert da quantidade

    const button = screen.getByTestId('carrinhoButton');

    fireEvent.press(button);

    expect(Alert.alert as jest.Mock).toHaveBeenCalledWith("Alerta",
      "Produto já adicionado?",
      [
        {
          text: "Cancelar",
          style: "cancel"
        },

      ]);


  });
  it("shows alert for lack of product quantity", async () => {


    const button = screen.getByTestId('carrinhoButton');

    fireEvent.press(button);


    expect(Alert.alert as jest.Mock).toHaveBeenCalledWith("Alerta",
      "Adicione uma quantidade de produto",
      [
        {
          text: "Cancelar",
          style: "cancel"
        },

      ]);

  });

  afterEach(() => {
    jest.restoreAllMocks();
    cleanup()
  });
});