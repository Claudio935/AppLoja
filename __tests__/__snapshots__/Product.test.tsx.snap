// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> renders correctly 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "marginBottom": 15,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "justifyContent": "space-between",
            "marginVertical": 15,
            "width": "100%",
          }
        }
      >
        <Image
          source={
            {
              "uri": "./img/testMasculino",
            }
          }
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "height": 220,
              "justifyContent": "center",
              "marginBottom": 15,
              "width": 180,
            }
          }
          testID="imageProduto"
        />
        <Text
          style={
            {
              "color": "#000",
              "fontSize": 15,
              "fontWeight": "900",
              "marginBottom": 15,
              "textAlign": "center",
              "width": 300,
            }
          }
        >
          testMasculino
        </Text>
        <Text
          style={
            {
              "marginBottom": 15,
              "paddingHorizontal": 5,
              "textAlign": "justify",
            }
          }
        >
          testandoMasculino
        </Text>
        <Text
          style={
            {
              "color": "#b71919",
              "fontSize": 14,
              "fontWeight": "900",
              "marginBottom": 15,
            }
          }
        >
          R$Â 0,00
        </Text>
        <View
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
              "marginBottom": 15,
              "width": "100%",
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="decrement"
          >
            <View
              style={
                [
                  {},
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                &lt;
              </Text>
            </View>
          </View>
          <Text
            style={
              {
                "fontWeight": "900",
                "marginHorizontal": 15,
              }
            }
            testID="textQuantidade"
          >
            0
          </Text>
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="increment"
          >
            <View
              style={
                [
                  {},
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                &gt;
              </Text>
            </View>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="carrinhoButton"
        >
          <View
            style={
              [
                {},
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Adicionar ao Carrinho
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
